### DISCUSSION TITLE
Two modifiers with one key

### DISCUSSION THREAD

**Discussion by donjan** [vial-kb/vial-gui#278] [Q&A]
I'm trying to set up a key to hold two modifiers at once.

More specifically: on layer 1, I sometimes want to modify keys with `Shift`. Currently, this requires holding both my `MO(1)` and `Shift` keys.
Instead, I would like to move to layer 1 with a simple `MO(1)` key, then slide over (with a tiny overlap period) to another key that holds `Shift` and `MO(1)` at the same time.
Can this be achieved?

On that second key, I've tried using a macro `[["down", "MO(0)", "KC_LSHIFT"]]` (also tried swapped order), but it doesn't work.

I've further tried assigning `LCTL(LSFT(KC))` as per https://www.reddit.com/r/qmk/comments/1aie4pu/hold_two_mods_with_one_key/, but the GUI won't accept that.

**Comment by xyzz** [comment]
I don't quite understand why the title is about two modifiers but in the description you're mostly talking about holding shift (except for the last point about LCTL(LSFT()) which seems disconnected from the rest. In any case, it seems what you want is `LM(layer, mod)` which is described here: https://github.com/qmk/qmk_firmware/blob/5dc92edf11946dca663c6723d8039e734126ccb0/docs/feature_layers.md#switching-and-toggling-layers-switching-and-toggling-layers

**Reply by donjan** [reply]
Thanks for your reply. Unfortunately I'm not quite seeing how to make that work.

To be more clear, on layer 0, I have this:
![Screenshot from 2024-12-06 13-49-50](https://github.com/user-attachments/assets/1b03246b-6078-4775-8ef3-ce25517390bf)
Which is as it should stay.

On layer 1, I currently have this:
![Screenshot from 2024-12-06 13-51-28](https://github.com/user-attachments/assets/e8018452-21a4-4088-adc4-25fa5b45f164)
Which I want to improve: to `Shift`-modify other layer-1 keys, I'd like to go from holding both these keys to sliding over and holding just the right one.
I've tried setting it to `LM(1, KC_LSHIFT)`, which results in the computed value `0x5011`:
![Screenshot from 2024-12-06 13-55-01](https://github.com/user-attachments/assets/039a6db4-fc35-48ac-bfd5-d6d54e52b250)
Now if I hold down both, for some reason this behaves as if `Ctrl` and `MO(1)` were held. If I let go of the left one and keep holding the right one, it behaves as if just `Ctrl` were held.
Is it even possible to let go of the `MO(1)` key but still keep the layer 1 functionality of the right key?

Side point about the "two modifiers": given that the modifiers are kind of layers, and layers semantically kind of modify keys, I've taken to calling all of them "modifiers". Which may not be the most precise language. That last link was just about making anything work at all.


EDIT:
According to the `LM` documentation, I've initially tried to use `MOD_LSHIFT` (because `..._LSHIFT` is given by the Vial GUI), which failed, but `MOD_LSFT` gets accepted. The computed keycode for `LM(1, MOD_LSFT)` is `0x5012`.

Now holding both gives the desired behaviour (shifting layer-1 keys), but letting go of the left key drops back to shifting layer-0 keys. Can I make that stick?

**Reply by donjan** [reply]
Going by the Vial GUI, I've tried the fixed-layer LT on layer 1:
![Screenshot from 2024-12-08 13-09-39](https://github.com/user-attachments/assets/23bc56fd-05f0-4dd3-9c7e-553f4cac2cea)

But this reverts to layer 0 as soon as I let go of the left key (after holding both). Interesting that it doesn't stay on layer 1... what am I misunderstanding here?
Also, it's just tapping `Shift` instead of holding it, so not actually what I need.

**Reply by donjan** [reply]
I guess here's my problem...
Layer 0:
![Screenshot from 2024-12-08 13-27-48](https://github.com/user-attachments/assets/b99630cb-5592-496b-ab94-cef2f51c4b1a)
Layer 1:
![Screenshot from 2024-12-08 13-27-53](https://github.com/user-attachments/assets/affad906-103d-46b1-a1c1-8f401a98f728)
Holding the left key, then holding the right key, then letting go of the left key doesn't stay on layer 1.

Is this intended?
And is there a way to achieve holding a layer? (this might be an xy-ask, I want to hold `Shift` + `MO(1)` with just a single key on layer 1)
